doc/src/manual/style-guide.md

```diff
@@ -119,6 +119,10 @@ with both copying and modifying forms (e.g., [`sort`](@ref) and [`sort!`](@ref))
 which are just modifying (e.g., [`push!`](@ref), [`pop!`](@ref), [`splice!`](@ref)).  It
 is typical for such functions to also return the modified array for convenience.
 
+Functions related to IO or making use of random number generators (RNG) are notable exceptions:
+Since these functions almost invariably must mutate the IO or RNG, respectively, functions ending with `!` are used to signify a mutation _other_ than mutating the IO or incrementing the RNG state.
+For example, `rand(x)` mutates the RNG, whereas `rand!(x)` mutates the RNG and `x`; similarly, `read(io)` mutates `io`, whereas `read!(io, x)` mutates both arguments.
```

Comment:
> Since these functions almost invariably must mutate the IO or RNG

The other special property they have is that the io/rng argument is optional. In a sense this choice makes `println(x)` the canonical form (does not mutate any of its arguments, although it alters your screen) while `println(io, x)` is for experts? Ditto `rand(n)` vs `rand(rng, n)`.

Reply:
Good observation. `close(io)` has no fallback to `close()`, though. Everything here as exceptions 😕

Reply:
introducing `close!(io)` sounds like a good idea to me

Reply:
If we replaced `close` with `close!`, then we would have (unless I'm missing something) "functions ending with `!` are used to signify a mutation _other_ than reading or writing to IO or advancing the RNG state by generating random data."

Random.seed!(4) is another exception to the "is the mutated variable optional" rule.

Reply:
> The other special property they have is that the io/rng argument is optional

I don't think that's true. Nearly all IO operations don't have the IO optional. Neither is it the case that the mutation is a kind of side-effect: For functions like `close` or `seek`, it's the only desired effect.

I've switched "increment" to "advance".

Reply:
Those are exceptions to the "functions ending with ! are used to signify a mutation other than reading or writing to IO or advancing the RNG state by generating random data." rule I proposed, not exceptions to what this PR says, right?

Reply:
> Nearly all IO operations

counting distinct functions? But counting by use in the wild, surely nearly all have optional IO.

For RNG, which ones don't make it optional?